File: 631 task sceduler/task_sceduler.py
import collections

class task_sceduler(object):
    def leastInterval1(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        task_counts = collections.Counter(tasks).values()
        Max_freq = max(task_counts)
        Max_freq_count = task_counts.count(Max_freq)
        return max(len(tasks), (Max_freq - 1) * (n + 1) + Max_freq_count)
    
    def leastInterval2(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        task_counts = collections.Counter(tasks).values()
        Max_freq = max(task_counts)
        Max_freq_count = task_counts.count(Max_freq)
        
        empty_slots = (n - Max_freq_count + 1) * (Max_freq - 1)
        available_tasks = len(tasks) - Max_freq *  Max_freq_count;
        
        idles = max(0, empty_slots - available_tasks);
        
        return len(tasks) + idles

File: 631 task sceduler/TaskSceduler.java
public class TaskSceduler {
    public int leastInterval(char[] tasks, int n) {
        int[] map = new int[26];
        for (char task : tasks) {
            map[task - 'A']++;
        }
        int max = 0;
        int maxCount = 0;
        for (int num : map) {
            if (max < num) {
                max = num;
                maxCount = 1;
            } else if (max == num) {
                maxCount++;
            }
        }
        
        int emptySlots = (n - maxCount + 1) * (max - 1);
        int availableTasks = tasks.length - max * maxCount;
        int idles = Math.max(0, emptySlots - availableTasks);
        
        return tasks.length + idles;
    }

    // explanation: https://leetcode.com/problems/task-scheduler/discuss/104500/Java-O(n)-time-O(1)-space-1-pass-no-sorting-solution-with-detailed-explanation
}
